Audio refactor

finneyj opened the pull request:
Not ready to merge. Raising this PR for review purposes only.

finneyj committed:
Remove runtime configuration of bitrate for StreamSplitter

  - Add support for dynamic reconfiguraton of drop rate
  - Remove methods and events associates with runtime reconfiguration
  - Fix minor bug in dissemination of data rate

finneyj committed:
Remove legacy code.

finneyj committed:
Introduce DataSourceSink base class

  - Refactor to have a common base class
  - Standardize implementation
  - reduce duplicted code

n.b. this is a first cut, compiling but untestsed code.

finneyj committed:
More refactoring.

Compiles, but not tested.

finneyj committed:
Bugfix minor errors

finneyj committed:
Add tri-state for connectedness of audio filters

  - add
  - bugfix

microbit-carlos commented on source/streams/StreamRecording.cpp:
```diff
@@ -251,11 +222,3 @@ bool StreamRecording::isStopped()
     return this->state == REC_STATE_STOPPED;
 }
-
-float StreamRecording::getSampleRate()
-{
-    if( this->lastUpstreamRate == DATASTREAM_SAMPLE_RATE_UNKNOWN )
-        return this->upStream.getSampleRate();
-    
-    return this->lastUpstreamRate;
-}
```
The `StreamRecording::getSampleRate()` method has been removed by the `virtual float getSampleRate()` declaration has been left in `StreamRecording.h`, which causes a compilation error.

finneyj committed:
Refactored stream recording class

  - simplified structure
  - removed old/redundant functions
  - utilised audio-refactor APIs
  - implemented copy-mode recording (as yet untested) for smaller blocks.

Known working with the following program:

int main()
{
    uBit.init();

    uBit.audio.requestActivation();

    SplitterChannel *channel = uBit.audio.splitter->createChannel();
    StreamRecording *recording = new StreamRecording(*channel);
    MixerChannel * mix = uBit.audio.mixer.addChannel(*recording);

    recording->record();
    mix->setSampleRate(11000);

    uBit.sleep(1000);

    while(1)
    {
        DMESG("PLAYING...");
        recording->play();
        DMESG("DONE.");
        uBit.sleep(10000);
    }
}

finneyj committed:
Bugfix for smaller packets being received from the microphone

  - remove excessive debug information
  - makeif statements positive, not negative

finneyj committed:
Enable back to back recordAsync / playbackAsync

finneyj committed:
Fix typo in test expression. classic!

finneyj committed:
Introduce (optional) StreamAnalyzer class

finneyj committed:
Update LevelDetectorSPL to always wake sleeping fibers.

finneyj committed:
Fix bug relating to reset of LevelDetectorSPL state

finneyj committed:
Add support for detecting removal of MessageBus listeners

finneyj committed:
Fix synchronous playback bug via SerialStreamer API (bug #477)

  - Fixed incorrent (float) cast position
  - Prevent locakups cause by synchronous playout interfaces

finneyj committed:
Introduce SerialStreamer class into codal-core

  - Update Serial base class to also hold a pointer to the first instance created
  - Update SerialStreamer class to use that instance as default.

finneyj committed:
Register SerialStreamer component as always wantingData

  - prevents bugs such as #479 in situatiosn where the stream uses auto-activation.

finneyj committed:
Remove stale data stored in DataStream

 - also code cleanup to remove unused code.

finneyj committed:
Add noise gate to level detector SPL

  - provides cleaner data outpu for users when using higher gain settings with noisy microphones.

finneyj committed:
Add optional support to filter unwanted packets at source

finneyj committed:
Create tailored dataWanted() behaviour for SerialStremaer (#485)

  - also remove superfluous return in DataSinkSource
